zenoh/src/api/selector.rs

```diff
@@ -60,157 +55,125 @@ use super::{key_expr::KeyExpr, queryable::Query};
-///
-/// Most methods will return an Error if duplicates of a same parameter are found, to avoid HTTP Parameter Pollution like vulnerabilities.
-#[repr(transparent)]
-#[derive(Clone, PartialEq, Eq)]
-pub struct Parameters<'a>(Properties<'a>);
-
-impl<'a> Deref for Parameters<'a> {
-    type Target = Properties<'a>;
-
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
 }
 
-impl<'a> DerefMut for Parameters<'a> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
-    }
-}
-
-impl std::fmt::Display for Parameters<'_> {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.0)
-    }
-}
-
-impl std::fmt::Debug for Parameters<'_> {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self)
-    }
-}
-
-impl<'a, T> From<T> for Parameters<'a>
+impl<'a, K, P> From<(K, P)> for Selector<'a>
 where
-    T: Into<Properties<'a>>,
+    K: Into<KeyExpr<'a>>,
+    P: Into<Parameters<'a>>,
 {
-    fn from(value: T) -> Self {
-        Parameters(value.into())
+    fn from((key_expr, parameters): (K, P)) -> Self {
+        Self::owned(key_expr, parameters)
     }
 }
 
-impl<'s> From<&'s Parameters<'s>> for HashMap<&'s str, &'s str> {
-    fn from(props: &'s Parameters<'s>) -> Self {
-        HashMap::from(&props.0)
+impl<'a> From<Selector<'a>> for (KeyExpr<'a>, Parameters<'a>) {
+    fn from(selector: Selector<'a>) -> Self {
+        (
+            selector.key_expr.into_owned(),
+            selector.parameters.into_owned(),
+        )
     }
 }
 
-impl From<&Parameters<'_>> for HashMap<String, String> {
-    fn from(props: &Parameters) -> Self {
-        HashMap::from(&props.0)
+impl<'a> From<&'a Selector<'a>> for (&'a KeyExpr<'a>, &'a Parameters<'a>) {
+    fn from(selector: &'a Selector<'a>) -> Self {
+        (selector.key_expr.as_ref(), selector.parameters.as_ref())
     }
 }
 
-impl From<Parameters<'_>> for HashMap<String, String> {
-    fn from(props: Parameters) -> Self {
-        HashMap::from(props.0)
-    }
+#[zenoh_macros::unstable]
+pub trait PredefinedParameters {
+    const REPLY_KEY_EXPR_ANY_SEL_PARAM: &'static str = "_anyke";
+    const TIME_RANGE_KEY: &'static str = "_time";
+    /// Sets the time range targeted by the selector parameters.
+    fn set_time_range<T: Into<Option<TimeRange>>>(&mut self, time_range: T);
+    /// Sets the parameter allowing to receieve replies from queryables not matching
+    /// the requested key expression. This may happen in this scenario:
+    /// - we are requesting keyexpr `a/b`.
+    /// - queryable is declared to handle `a/*` requests and contains data for `a/b` and `a/c`. 
+    /// - queryable receives our request and sends two replies with data for `a/b` and `a/c`
+    /// 
+    /// Normally only `a/b` reply would be accepted, but with `_anyke` parameter set, both replies are accepted.
+    fn set_reply_key_expr_any(&mut self);
```

Comment:
@Mallets could you please check if I understand the logic of `_anyke` parameter correctly?

Reply:
What you wrote is correct. In addition to that, when `_anyke` is defined, literally *any key expression* is allowed in the reply. I.e., the queryable could reply with `x/y/z` and that will be accepted.